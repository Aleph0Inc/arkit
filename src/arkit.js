"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const logger_1 = require("./logger");
const config_1 = require("./config");
const parser_1 = require("./parser");
const generator_1 = require("./generator");
exports.arkit = (directory) => {
    const config = new config_1.Config(directory);
    logger_1.debug(`Config`);
    logger_1.trace(config);
    const parser = new parser_1.Parser(config);
    const files = parser.parse();
    logger_1.trace('Parsed files');
    logger_1.trace(files);
    const generator = new generator_1.Generator(config, files);
    return Promise.all(config.outputs.map(output => {
        const puml = generator.generatePlantUML(output);
        if (output.path) {
            for (const outputPath of config.array(output.path)) {
                const fullExportPath = path.join(config.directory, outputPath);
                const ext = path.extname(fullExportPath);
                if (fs.existsSync(fullExportPath)) {
                    fs.unlinkSync(fullExportPath);
                }
                if (ext === '.puml') {
                    fs.writeFileSync(fullExportPath, puml);
                }
                if (ext === '.svg') {
                    generator.convertToSVG(puml).then(svg => {
                        fs.writeFileSync(fullExportPath, svg);
                    }).catch(err => {
                        throw err;
                    });
                }
            }
            return puml;
        }
        return generator.convertToSVG(puml).then(svg => {
            console.log(svg);
            return puml;
        }).catch(err => {
            throw err;
        });
    }));
};
