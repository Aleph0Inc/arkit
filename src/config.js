"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const logger_1 = require("./logger");
/**
 * Component name formats
 */
var ComponentNameFormat;
(function (ComponentNameFormat) {
    ComponentNameFormat["BASE_NAME"] = "base";
    ComponentNameFormat["FULL_NAME"] = "full";
    ComponentNameFormat["COMPLETE_PATH"] = "complete";
})(ComponentNameFormat = exports.ComponentNameFormat || (exports.ComponentNameFormat = {}));
var OutputDirection;
(function (OutputDirection) {
    OutputDirection["HORIZONTAL"] = "horizontal";
    OutputDirection["VERTICAL"] = "vertical";
})(OutputDirection = exports.OutputDirection || (exports.OutputDirection = {}));
const DEFAULT_COMPONENTS = {
    type: 'Component',
    patterns: ['**/*.ts', '**/*.js', '**/*.jsx', '**/*.tsx']
};
exports.DEFAULT_CONFIG = {
    components: DEFAULT_COMPONENTS,
    excludePatterns: ['node_modules/**', 'test/**', '**/*.test.*', '**/*.spec.*'],
    output: {}
};
class Config {
    constructor(directory) {
        this.directory = directory;
        this.patterns = [];
        this.excludePatterns = [];
        this.extensions = ['.js', '.ts', '.jsx', '.tsx'];
        logger_1.info('Working directory', directory);
        const userConfigPath = path.join(this.directory, 'arkit');
        const userConfig = this.safeRequire(userConfigPath);
        this.components = this.array(userConfig && userConfig.components) || [];
        if (!this.components.length) {
            this.components.push(DEFAULT_COMPONENTS);
        }
        this.outputs = this.array(userConfig && userConfig.output || exports.DEFAULT_CONFIG.output);
        if (userConfig && userConfig.excludePatterns) {
            this.excludePatterns.push(...userConfig.excludePatterns);
        }
        else if (!userConfig && exports.DEFAULT_CONFIG.excludePatterns) {
            this.excludePatterns.push(...exports.DEFAULT_CONFIG.excludePatterns);
        }
        for (const component of this.components) {
            if (component.patterns) {
                this.patterns.push(...component.patterns);
            }
            if (component.excludePatterns) {
                this.excludePatterns.push(...component.excludePatterns);
            }
        }
    }
    safeRequire(path) {
        try {
            return require(path);
        }
        catch (e) {
            logger_1.warn(e.toString());
        }
    }
    array(input) {
        if (input) {
            return [].concat(input);
        }
    }
}
exports.Config = Config;
