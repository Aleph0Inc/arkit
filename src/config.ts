import * as path from 'path'
import { ComponentSchema, ConfigSchema, Options, OutputSchema } from './schema'
import { array, safeRequire } from './utils'

const DEFAULT_COMPONENTS: ComponentSchema[] = [
  {
    type: 'Component',
    patterns: ['**/*.ts', '**/*.js', '**/*.jsx', '**/*.tsx']
  },
  {
    type: 'Dependency',
    patterns: ['node_modules/*']
  }
]

export class Config {
  directory: string
  components: ComponentSchema[]
  outputs: OutputSchema[]
  patterns: string[] = []
  excludePatterns: string[]
  extensions = ['.js', '.ts', '.jsx', '.tsx']

  constructor (options: Options) {
    this.directory = options.directory
    const userConfigPath = path.resolve(this.directory, 'arkit')
    const userConfig = safeRequire<ConfigSchema>(userConfigPath)

    this.components = array(userConfig && userConfig.components) || []

    if (!this.components.length) {
      this.components.push(...DEFAULT_COMPONENTS)
    }

    this.outputs = this.getOutputs(options, userConfig)
    this.excludePatterns = this.getExcludePatterns(options, userConfig)

    for (const component of this.components) {
      if (component.patterns) {
        this.patterns.push(...component.patterns)
      }
    }
  }

  private getOutputs (options: Options, userConfig?: ConfigSchema): OutputSchema[] {
    const userConfigOutput = userConfig && userConfig.output
    const outputOption = options.output && options.output.length ? options.output : undefined
    const firstOption = options.first && options.first.length ? options.first : undefined
    const shouldGenerateOutput = outputOption || firstOption || !userConfigOutput

    if (!shouldGenerateOutput) {
      return array(userConfigOutput)!
    }

    const firstGroup = firstOption ? [{
      first: true,
      patterns: firstOption
    }] : []

    return [
      {
        path: outputOption,
        groups: [
          ...firstGroup,
          { type: 'Dependencies', components: ['Dependency'] },
          {} // everything else
        ]
      }
    ]
  }

  private getExcludePatterns (options: Options, userConfig?: ConfigSchema): string[] {
    const excludePatterns: string[] = []

    if (options.exclude) {
      excludePatterns.push(...options.exclude)
    }

    if (userConfig && userConfig.excludePatterns) {
      excludePatterns.push(...userConfig.excludePatterns)
    }

    for (const component of this.components) {
      if (component.excludePatterns) {
        excludePatterns.push(...component.excludePatterns)
      }
    }

    return excludePatterns
  }
}
