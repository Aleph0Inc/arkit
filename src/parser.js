"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const nanomatch = require("nanomatch");
const os_1 = require("os");
const ts_morph_1 = require("ts-morph");
const resolve_1 = require("resolve");
const logger_1 = require("./logger");
const tsconfig_paths_1 = require("tsconfig-paths");
const QUOTES = `(?:'|")`;
const TEXT_INSIDE_QUOTES = `${QUOTES}([^'"]+)${QUOTES}`;
const TEXT_INSIDE_QUOTES_RE = new RegExp(TEXT_INSIDE_QUOTES);
const REQUIRE_RE = new RegExp(`require\\(${TEXT_INSIDE_QUOTES}\\)(?:\\.(\\w+))?`);
class Parser {
    constructor(config) {
        this.config = config;
        this.sourceFiles = new Map();
        this.tsResolutionCache = new Map();
        this.resolveTsConfigPaths();
        this.project = new ts_morph_1.Project({
            tsConfigFilePath: this.tsConfigFilePath,
            addFilesFromTsConfig: false
        });
        logger_1.debug('Adding directory...', this.config.directory);
        this.project.addExistingDirectory(this.config.directory);
        logger_1.debug('Searching files...');
        const allFilePaths = this.walkSync(this.config.directory)
            .map(filepath => path.relative(this.config.directory, filepath));
        const suitableFilePaths = nanomatch(allFilePaths, this.config.patterns, { ignore: this.config.excludePatterns })
            .map(filepath => path.join(this.config.directory, filepath));
        logger_1.trace(suitableFilePaths);
        logger_1.debug(`Adding ${suitableFilePaths.length} files...`);
        this.project.addExistingSourceFiles(suitableFilePaths).forEach(sourceFile => {
            this.sourceFiles.set(sourceFile.getFilePath(), sourceFile);
        });
    }
    resolveTsConfigPaths() {
        const tsConfig = tsconfig_paths_1.loadConfig(this.config.directory);
        if (tsConfig.resultType === 'success') {
            this.tsConfigFilePath = path.relative(this.config.directory, tsConfig.configFileAbsolutePath);
            logger_1.debug('Found TypeScript config', this.tsConfigFilePath);
            logger_1.debug('Registering ts-config paths...');
            this.tsResolve = tsconfig_paths_1.createMatchPath(tsConfig.absoluteBaseUrl, tsConfig.paths, tsConfig.mainFields, tsConfig.addMatchAll);
            logger_1.debug(tsConfig.paths);
        }
    }
    parse() {
        logger_1.debug('Parsing...', this.sourceFiles.size, 'files');
        const files = {};
        for (const [fullPath, sourceFile] of this.sourceFiles) {
            const filePath = path.relative(this.config.directory, fullPath);
            const statements = sourceFile.getStatements();
            logger_1.debug(filePath, statements.length, 'statements');
            const exports = this.getExports(sourceFile, statements);
            const imports = this.getImports(sourceFile, statements);
            logger_1.debug(Object.keys(exports).length, 'exports', Object.keys(imports).length, 'imports');
            files[filePath] = { exports, imports };
        }
        return files;
    }
    getImports(sourceFile, statements) {
        return statements.reduce((imports, statement) => {
            let sourceFileImports;
            let structure;
            if (ts_morph_1.TypeGuards.isVariableStatement(statement)) {
                statement.getStructure().declarations.forEach(declaration => {
                    if (typeof declaration.initializer === 'string') {
                        const [match, moduleSpecifier, namedImport] = Array.from(REQUIRE_RE.exec(declaration.initializer) || []);
                        if (moduleSpecifier) {
                            const importedFile = this.resolveModule(moduleSpecifier, sourceFile);
                            sourceFileImports = this.addImportedFile(importedFile, imports);
                            if (sourceFileImports && namedImport) {
                                sourceFileImports.push(namedImport);
                            }
                        }
                    }
                });
            }
            if (ts_morph_1.TypeGuards.isImportDeclaration(statement) || ts_morph_1.TypeGuards.isExportDeclaration(statement)) {
                let moduleSpecifier;
                try {
                    structure = statement.getStructure();
                    moduleSpecifier = structure.moduleSpecifier;
                }
                catch (e) {
                    logger_1.warn(e);
                    const brokenLineNumber = statement.getStartLineNumber();
                    const brokenLine = sourceFile.getFullText().split(os_1.EOL)[brokenLineNumber - 1];
                    const moduleSpecifierMatch = TEXT_INSIDE_QUOTES_RE.exec(brokenLine);
                    if (moduleSpecifierMatch) {
                        moduleSpecifier = moduleSpecifierMatch[1];
                    }
                }
                if (moduleSpecifier) {
                    const importedFile = (structure && statement.getModuleSpecifierSourceFile()) ||
                        this.resolveModule(moduleSpecifier, sourceFile);
                    sourceFileImports = this.addImportedFile(importedFile, imports);
                }
            }
            if (ts_morph_1.TypeGuards.isImportDeclaration(statement) && sourceFileImports && structure) {
                const importStructure = structure;
                if (importStructure.namespaceImport) {
                    sourceFileImports.push(importStructure.namespaceImport);
                }
                if (importStructure.defaultImport) {
                    sourceFileImports.push(importStructure.defaultImport);
                }
                if (importStructure.namedImports instanceof Array) {
                    sourceFileImports.push(...importStructure.namedImports.map(namedImport => typeof namedImport === 'string' ? namedImport : namedImport.name));
                }
                if (!sourceFileImports.length && !importStructure.namedImports) {
                    logger_1.warn('IMPORT', sourceFile.getBaseName(), structure);
                }
            }
            return imports;
        }, {});
    }
    addImportedFile(importedFile, imports) {
        if (importedFile) {
            const filePath = path.relative(this.config.directory, importedFile.getFilePath());
            return imports[filePath] = imports[filePath] || [];
        }
    }
    getExports(sourceFile, statements) {
        return statements.reduce((exports, statement) => {
            if (ts_morph_1.TypeGuards.isExportableNode(statement) && statement.hasExportKeyword()) {
                if (ts_morph_1.TypeGuards.isVariableStatement(statement)) {
                    const structure = statement.getStructure();
                    exports = [
                        ...exports,
                        structure.declarations.map(declaration => declaration.name)
                    ];
                }
                else if (ts_morph_1.TypeGuards.isInterfaceDeclaration(statement) ||
                    ts_morph_1.TypeGuards.isClassDeclaration(statement) ||
                    ts_morph_1.TypeGuards.isEnumDeclaration(statement) ||
                    ts_morph_1.TypeGuards.isFunctionDeclaration(statement) ||
                    ts_morph_1.TypeGuards.isTypeAliasDeclaration(statement)) {
                    logger_1.trace('EXPORT', sourceFile.getBaseName(), statement.getStructure());
                }
                else {
                    logger_1.warn('EXPORT Unknown type', sourceFile.getBaseName(), statement);
                }
            }
            return exports;
        }, []);
    }
    resolveModule(moduleSpecifier, sourceFile) {
        let modulePath;
        if (!this.tsResolutionCache.has(moduleSpecifier)) {
            try {
                modulePath = resolve_1.sync(moduleSpecifier, {
                    basedir: sourceFile.getDirectoryPath(),
                    extensions: this.config.extensions
                });
            }
            catch (e) {
                modulePath = this.resolveTsModule(moduleSpecifier);
                this.tsResolutionCache.set(moduleSpecifier, modulePath);
            }
        }
        else {
            modulePath = this.tsResolutionCache.get(moduleSpecifier);
        }
        if (modulePath) {
            return this.sourceFiles.get(modulePath);
        }
        else {
            logger_1.trace('Import not found', sourceFile.getBaseName(), moduleSpecifier);
        }
    }
    resolveTsModule(moduleSpecifier) {
        if (!this.tsResolve)
            return;
        logger_1.trace('Resolve TS', moduleSpecifier);
        const modulePath = this.tsResolve(moduleSpecifier);
        if (!modulePath)
            return;
        for (const ext of this.config.extensions) {
            const fullPath = `${modulePath}${ext}`;
            if (this.sourceFiles.has(fullPath)) {
                return fullPath;
            }
        }
    }
    walkSync(p) {
        if (fs.statSync(p).isDirectory()) {
            return fs.readdirSync(p).reduce((paths, f) => [...paths, ...this.walkSync(path.join(p, f))], []);
        }
        else {
            return [p];
        }
    }
}
exports.Parser = Parser;
