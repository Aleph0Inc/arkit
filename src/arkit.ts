import * as fs from 'fs'
import * as path from 'path'
import { debug, trace } from './logger'
import { Config } from './config'
import { Parser } from './parser'
import { Generator } from './generator'

export interface Options {
  directory: string,
  output?: string[],
  first?: string[]
}

const getOptions = (options?: Options): Options => {
  const opts: Options = {
    directory: (options && options.directory) || '',
    output: (options && options.output) || [],
    first: (options && options.first) || []
  }
  const directory = path.isAbsolute(opts.directory) ? opts.directory : path.join(process.cwd(), opts.directory)

  return {
    ...opts,
    directory
  }
}

export const arkit = (options?: Options): Promise<string[]> => {
  const opts = getOptions(options)
  debug('Options')
  debug(opts)

  const config = new Config(opts)

  debug('Config')
  debug(config)

  const parser = new Parser(config)
  const files = parser.parse()

  trace('Parsed files')
  trace(files)

  const generator = new Generator(config, files)

  return Promise.all(config.outputs.reduce((promises, output) => {
    const puml = generator.generatePlantUML(output)

    if (output.path && output.path.length) {
      for (const outputPath of config.array(output.path)!) {
        promises.push(generator.convert(outputPath, puml))
      }
    } else {
      promises.push(generator.convert('svg', puml))
    }

    return promises
  }, [] as Promise<string>[]))
}
