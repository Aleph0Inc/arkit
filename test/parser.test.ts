import * as path from 'path'
import { Config } from '../src/config'
import { Parser } from '../src/parser'

describe('Parser', () => {
  test('Arkit', () => {
    const directory = path.resolve(__dirname, '../src')
    const parser = new Parser(new Config({directory}))

    expect(parser.parse()).toMatchSnapshot()
  })

  test('Sample', () => {
    const directory = path.resolve(__dirname, './sample')
    const config = new Config({directory})
    const parser = new Parser(config)

    expect(parser.parse()).toMatchSnapshot()
  })

  test('Angular2 Todo', () => {
    const directory = path.resolve(__dirname, './angular2_es2015')
    const parser = new Parser(new Config({directory}))

    expect(parser.parse()).toMatchSnapshot()
  })

  test('Express', () => {
    const directory = path.resolve(__dirname, './express')
    const parser = new Parser(new Config({directory}))

    expect(parser.parse()).toMatchSnapshot()
  })
})
