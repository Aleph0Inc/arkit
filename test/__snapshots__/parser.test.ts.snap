// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser Angular2 Todo 1`] = `
Object {
  "app/components/app/app.component.js": Object {
    "exports": Array [
      "AppComponent",
    ],
    "imports": Object {
      "app/components/app/app.template.html": Array [
        "template",
      ],
    },
  },
  "app/components/index.js": Object {
    "exports": Array [],
    "imports": Object {
      "app/components/app/app.component.js": Array [],
      "app/components/todo-footer/todo-footer.component.js": Array [],
      "app/components/todo-header/todo-header.component.js": Array [],
      "app/components/todo-item/todo-item.component.js": Array [],
      "app/components/todo-list/todo-list.component.js": Array [],
    },
  },
  "app/components/todo-footer/todo-footer.component.js": Object {
    "exports": Array [
      "TodoFooterComponent",
    ],
    "imports": Object {
      "app/components/todo-footer/todo-footer.template.html": Array [
        "template",
      ],
      "app/services/todo-store.service.js": Array [
        "TodoStoreService",
      ],
    },
  },
  "app/components/todo-header/todo-header.component.js": Object {
    "exports": Array [
      "TodoHeaderComponent",
    ],
    "imports": Object {
      "app/components/todo-header/todo-header.template.html": Array [
        "template",
      ],
      "app/services/todo-store.service.js": Array [
        "TodoStoreService",
      ],
    },
  },
  "app/components/todo-item/todo-item.component.js": Object {
    "exports": Array [
      "TodoItemComponent",
    ],
    "imports": Object {
      "app/components/todo-item/todo-item.template.html": Array [
        "template",
      ],
    },
  },
  "app/components/todo-list/todo-list.component.js": Object {
    "exports": Array [
      "TodoListComponent",
    ],
    "imports": Object {
      "app/components/todo-list/todo-list.template.html": Array [
        "template",
      ],
      "app/services/todo-store.service.js": Array [
        "TodoStoreService",
      ],
    },
  },
  "app/main.js": Object {
    "exports": Array [],
    "imports": Object {
      "app/main.module.js": Array [
        "MainModule",
      ],
      "node_modules/rxjs/index.js": Array [],
    },
  },
  "app/main.module.js": Object {
    "exports": Array [
      "MainModule",
    ],
    "imports": Object {
      "app/components/index.js": Array [
        "AppComponent",
        "TodoListComponent",
        "TodoFooterComponent",
        "TodoHeaderComponent",
        "TodoItemComponent",
      ],
      "app/pipes/index.js": Array [
        "TrimPipe",
      ],
      "app/routes.js": Array [
        "routes",
      ],
      "app/services/todo-store.service.js": Array [
        "TodoStoreService",
      ],
    },
  },
  "app/models/todo.model.js": Object {
    "exports": Array [
      "TodoModel",
    ],
    "imports": Object {},
  },
  "app/pipes/index.js": Object {
    "exports": Array [],
    "imports": Object {
      "app/pipes/trim/trim.pipe.js": Array [],
    },
  },
  "app/pipes/trim/trim.pipe.js": Object {
    "exports": Array [
      "TrimPipe",
    ],
    "imports": Object {},
  },
  "app/routes.js": Object {
    "exports": Array [
      "routes",
    ],
    "imports": Object {
      "app/components/todo-list/todo-list.component.js": Array [
        "TodoListComponent",
      ],
    },
  },
  "app/services/todo-store.service.js": Object {
    "exports": Array [
      "TodoStoreService",
    ],
    "imports": Object {
      "app/models/todo.model.js": Array [
        "TodoModel",
      ],
    },
  },
  "main.bundle.js": Object {
    "exports": Array [],
    "imports": Object {},
  },
  "webpack.config.js": Object {
    "exports": Array [],
    "imports": Object {},
  },
}
`;

exports[`Parser Arkit 1`] = `
Object {
  "arkit.ts": Object {
    "exports": Array [
      "getConfig",
      "getOutputs",
      "arkit",
    ],
    "imports": Object {
      "config.ts": Array [
        "Config",
      ],
      "generator.ts": Array [
        "Generator",
      ],
      "node_modules/yargs/index.js": Array [
        "yargs",
      ],
      "parser.ts": Array [
        "Parser",
      ],
      "path": Array [
        "path",
      ],
      "types.ts": Array [
        "Options",
        "OutputFormat",
        "SavedString",
      ],
      "utils.ts": Array [
        "info",
        "trace",
      ],
    },
  },
  "config.ts": Object {
    "exports": Array [
      "Config",
    ],
    "imports": Object {
      "path": Array [
        "path",
      ],
      "types.ts": Array [
        "ComponentNameFormat",
        "ComponentSchema",
        "ConfigBase",
        "ConfigSchema",
        "GroupSchema",
        "Options",
        "OutputSchema",
      ],
      "utils.ts": Array [
        "array",
        "debug",
        "safeRequire",
      ],
    },
  },
  "generator.base.ts": Object {
    "exports": Array [
      "GeneratorBase",
    ],
    "imports": Object {
      "path": Array [
        "path",
      ],
      "types.ts": Array [
        "ComponentFilters",
        "ComponentNameFormat",
        "ComponentSchema",
        "OutputSchema",
        "Files",
        "Component",
        "Components",
        "EMPTY_LAYER",
        "Layers",
        "ConfigBase",
      ],
      "utils.ts": Array [
        "trace",
        "warn",
        "array",
        "match",
        "verifyComponentFilters",
      ],
    },
  },
  "generator.ts": Object {
    "exports": Array [
      "Generator",
    ],
    "imports": Object {
      "fs": Array [
        "fs",
      ],
      "generator.base.ts": Array [
        "GeneratorBase",
      ],
      "node_modules/progress/index.js": Array [
        "ProgressBar",
      ],
      "path": Array [
        "path",
      ],
      "types.ts": Array [
        "Component",
        "Context",
        "EMPTY_LAYER",
        "OutputDirection",
        "OutputFormat",
        "OutputSchema",
        "SavedString",
      ],
      "utils.ts": Array [
        "array",
        "bold",
        "debug",
        "getAllComponents",
        "info",
        "request",
        "trace",
      ],
    },
  },
  "logger.ts": Object {
    "exports": Array [
      "fatal",
      "error",
      "warn",
      "info",
      "debug",
      "trace",
    ],
    "imports": Object {
      "node_modules/pino/pino.js": Array [
        "pino",
      ],
    },
  },
  "parser.ts": Object {
    "exports": Array [
      "Parser",
    ],
    "imports": Object {
      "node_modules/progress/index.js": Array [
        "ProgressBar",
      ],
      "node_modules/resolve/index.js": Array [
        "sync",
      ],
      "node_modules/ts-morph/dist/main.js": Array [
        "ExportDeclarationStructure",
        "ImportDeclarationStructure",
        "Project",
        "SourceFile",
        "Statement",
        "TypeGuards",
      ],
      "node_modules/tsconfig-paths/lib/index.js": Array [
        "createMatchPath",
        "loadConfig",
        "MatchPath",
      ],
      "os": Array [
        "EOL",
      ],
      "path": Array [
        "path",
      ],
      "types.ts": Array [
        "ComponentSchema",
        "ConfigBase",
        "Exports",
        "Files",
        "Imports",
      ],
      "utils.ts": Array [
        "find",
        "getPaths",
        "debug",
        "info",
        "trace",
        "warn",
      ],
    },
  },
  "schema.ts": Object {
    "exports": Array [
      "Options",
      "ConfigSchema",
      "ComponentSchema",
      "ComponentNameFormat",
      "OutputSchema",
      "GroupSchema",
      "ComponentFilters",
      "OutputDirection",
      "OutputFormat",
    ],
    "imports": Object {},
  },
  "types.ts": Object {
    "exports": Array [
      "EMPTY_LAYER",
      "Components",
      "Layers",
      "Context",
      "Component",
      "Imports",
      "Exports",
      "File",
      "Files",
      "ConfigBase",
      "SavedString",
    ],
    "imports": Object {
      "schema.ts": Array [
        "ConfigSchema",
      ],
    },
  },
  "utils.ts": Object {
    "exports": Array [
      "getStats",
      "getMemoryUsage",
      "getPaths",
      "match",
      "find",
      "safeRequire",
      "array",
      "verifyComponentFilters",
      "bold",
      "request",
      "getAllComponents",
    ],
    "imports": Object {
      "fs": Array [
        "fs",
      ],
      "https": Array [
        "https",
      ],
      "logger.ts": Array [
        "trace",
        "warn",
      ],
      "node_modules/nanomatch/index.js": Array [
        "nanomatch",
      ],
      "path": Array [
        "path",
      ],
      "types.ts": Array [
        "Component",
        "ComponentFilters",
        "ComponentSchema",
        "Layers",
      ],
    },
  },
}
`;

exports[`Parser Express 1`] = `
Object {
  "application.js": Object {
    "exports": Array [],
    "imports": Object {
      "http": Array [],
      "middleware/init.js": Array [],
      "middleware/query.js": Array [],
      "node_modules/debug/src/index.js": Array [],
      "path": Array [
        "resolve",
      ],
      "router/index.js": Array [],
      "utils.js": Array [
        "compileETag",
        "compileQueryParser",
        "compileTrust",
      ],
      "view.js": Array [],
    },
  },
  "express.js": Object {
    "exports": Array [],
    "imports": Object {
      "application.js": Array [],
      "events": Array [
        "EventEmitter",
      ],
      "middleware/query.js": Array [],
      "request.js": Array [],
      "response.js": Array [],
      "router/index.js": Array [],
      "router/route.js": Array [],
    },
  },
  "middleware/init.js": Object {
    "exports": Array [],
    "imports": Object {},
  },
  "middleware/query.js": Object {
    "exports": Array [],
    "imports": Object {
      "node_modules/qs/lib/index.js": Array [],
    },
  },
  "request.js": Object {
    "exports": Array [],
    "imports": Object {
      "http": Array [],
      "net": Array [
        "isIP",
      ],
    },
  },
  "response.js": Object {
    "exports": Array [],
    "imports": Object {
      "http": Array [],
      "node_modules/safe-buffer/index.js": Array [
        "Buffer",
      ],
      "path": Array [],
      "utils.js": Array [
        "isAbsolute",
        "normalizeType",
        "normalizeTypes",
        "setCharset",
      ],
    },
  },
  "router/index.js": Object {
    "exports": Array [],
    "imports": Object {
      "node_modules/debug/src/index.js": Array [],
      "router/layer.js": Array [],
      "router/route.js": Array [],
    },
  },
  "router/layer.js": Object {
    "exports": Array [],
    "imports": Object {
      "node_modules/debug/src/index.js": Array [],
    },
  },
  "router/route.js": Object {
    "exports": Array [],
    "imports": Object {
      "node_modules/debug/src/index.js": Array [],
      "router/layer.js": Array [],
    },
  },
  "utils.js": Object {
    "exports": Array [],
    "imports": Object {
      "node_modules/qs/lib/index.js": Array [],
      "node_modules/safe-buffer/index.js": Array [
        "Buffer",
      ],
      "querystring": Array [],
    },
  },
  "view.js": Object {
    "exports": Array [],
    "imports": Object {
      "fs": Array [],
      "node_modules/debug/src/index.js": Array [],
      "path": Array [],
    },
  },
}
`;

exports[`Parser Sample 1`] = `
Object {
  "controllers/singleton.ts": Object {
    "exports": Array [
      "singleton",
    ],
    "imports": Object {
      "models/model.ts": Array [
        "List",
        "Model",
        "CONSTANT",
        "a",
        "b1",
        "b2",
        "b3",
        "c",
      ],
    },
  },
  "main.ts": Object {
    "exports": Array [],
    "imports": Object {
      "views/main.view.js": Array [
        "mainView",
      ],
    },
  },
  "models/model.ts": Object {
    "exports": Array [
      "Model",
      "List",
      "CONSTANT",
      "{ a, b: [b1, b2, b3], c }",
    ],
    "imports": Object {},
  },
  "views/base.component.js": Object {
    "exports": Array [],
    "imports": Object {},
  },
  "views/component.js": Object {
    "exports": Array [
      "Component",
    ],
    "imports": Object {
      "views/base.component.js": Array [
        "BaseComponent",
      ],
    },
  },
  "views/main.view.js": Object {
    "exports": Array [
      "mainView",
    ],
    "imports": Object {
      "controllers/singleton.ts": Array [
        "singleton",
      ],
      "views/component.js": Array [
        "Component",
      ],
    },
  },
}
`;
